### PR TITLE
Daria/thread names

### PR DESCRIPTION
**Pull request by dariagrudzien**
Welp! I got stuck trying to run a basic test o.O 

Please disregard the actual meat of the thing. This is my first time with Pytest and parameters, and I'm surprised that I don't see the new test in the report. I checked the `pyproject.toml` for config and I don't understand why this wouldn't run.

Also for fun when I started at least I was seeing `run-last-failure: 4 known failures not in selected tests`. Now in pycache directories I see files with the `.pyc` extension hinting at some assertion errors. When I try to run the specific test file, Pytest doesn't collect any items.

What am I missing?

```
╰─$ poetry run pytest --verbose
======================================================================================= test session starts ========================================================================================
platform darwin -- Python 3.11.4, pytest-7.4.0, pluggy-1.0.0 -- /Users/d/Library/Caches/pypoetry/virtualenvs/juniorguru-chick-woDL3_j3-py3.11/bin/python
cachedir: .pytest_cache
rootdir: /Users/d/dev/juniorguru/juniorguru-chick
configfile: pyproject.toml
testpaths: .
collected 3 items
run-last-failure: no previously failed tests, not deselecting items.

tests/test_lib_intro.py::test_choose_intro_emojis PASSED                                                                                                                                     [ 33%]
tests/test_lib_intro.py::test_choose_intro_emoji_edge_cases[z\xe1kladn\xed struktury v pythonu a C# v ITnetwork-<:csharp:842666113230045224>] PASSED                                         [ 66%]
tests/test_lib_intro.py::test_choose_intro_emoji_edge_cases[L\xe1k\xe1 m\u011b C++ a C#-<:cpp:842666129071931433>] PASSED                                                                    [100%]

========================================================================================= warnings summary =========================================================================================
../../../Library/Caches/pypoetry/virtualenvs/juniorguru-chick-woDL3_j3-py3.11/lib/python3.11/site-packages/discord/player.py:28
  /Users/d/Library/Caches/pypoetry/virtualenvs/juniorguru-chick-woDL3_j3-py3.11/lib/python3.11/site-packages/discord/player.py:28: DeprecationWarning: 'audioop' is deprecated and slated for removal in Python 3.13
    import audioop

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=================================================================================== 3 passed, 1 warning in 0.10s ===================================================================================
╭─
```

### CODE REVIEW

**Comment by honzajavorek** [comment]
I’m sorry for the lost hours! 😀 Easy to overlook. Im happy I could provide
second eyes. This is one of the things which are easy to spot on GitHub,
but I’m sure I could end up with the same problem myself!

File: tests/test_lib_thread_names.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import pytest
+from juniorguru_chick.lib.thread_names import name_thread
+
+@pytest.mark.parametrize('message, expected_name', [("[eslint, nextjs]", "eslint, nextjs"), ("[eslint,nextjs]", "eslint,nextjs"), ("[ Java ]", "Java"), ("[:css: CSS]", "CSS"), ("[eslint,nextjs, :css: CSS, ruby]", "eslint, nextjs, CSS, ruby")])
```
Comment (honzajavorek) [inline tests/test_lib_thread_names.py:4]: You probably want to have one per line - not per code style, but per having a good overview as a human.

File: juniorguru_chick/lib/thread_names.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import re
+
+def name_thread(message: str) -> str | None:
+    """Returns a name for a thread created from a given message. Example messages include: [eslint, nextjs], [eslint,nextjs], [ Java ], [:css:], [:css: CSS], [eslint,nextjs, :css: CSS]. Expected output includes: eslint, nextjs, eslint,nextjs, Java, css, CSS, eslint,nextjs, CSS."""
+
+    """
+    1. Check if the message includes text in brackets, if yes:
+        a. Remove brackets
+        b. Remove spaces
+        c. Check if the message includes text between colons
```
Comment (honzajavorek) [inline juniorguru_chick/lib/thread_names.py:10]: I think there will never be text between colons just like that. User types it there like `:css:`, but Discord translates it into either:

- Custom emoji, and it becomes `<:css:842343369618751519>`
- Unicode emoji, and it becomes `🐍`

For simplicity, I'd keep them as they are. Treat them as if they're any other words, and if that becomes a problem, let's fix that in the future. In such case, I think it should be enough to:

- See if the beginning (leading whitespace not breaking the algorithm) of the message has a match for `[`, followed by multiple "anything else than `]`", followed by `]`.
- If yes, take the "anything else than `]`", split it by commas, strip trailing white space.
- Join the resulting list back to a string and turn it into a thread name as you like.
- Keep the original message as is, we're just naming threads.
- If there are no brackets at the beginning, name the thread by the current weekday.

Potentially useful links:

- https://regex101.com/
- `str.lstrip()`, `re.match()`, `str.strip()`, `str.join()`
- Python's [verbose regular expressions](https://docs.python.org/3/library/re.html#re.VERBOSE) allow for less headaches, see e.g. [this](https://github.com/honzajavorek/junior.guru/blob/e8688cdb266d11353841343aa856ec1b49b5d9e2/juniorguru/lib/coupons.py#L4)

Also, the dispute whether to add [walrus operator](https://docs.python.org/3/whatsnew/3.8.html#assignment-expressions) to Python or not did cause Guido to step down as a BDFL, but in the end we have it! Checking results of `re.match` and `re.search` can be one line shorter now: `if match := re.match(...):` Ofc, I understand it's not for everyone, so up to you. (Just for those brave enough to roll on the beach with walruses, muhaha).

File: juniorguru_chick/lib/threads.py [hunk 3]
```diff
@@ -23,20 +23,21 @@ async def fetch_starting_message(thread: discord.Thread) -> discord.Message | No
     except discord.errors.NotFound:
         return None
 
+def name_thread(message: discord.Message, name_template) -> str | None:
```
Comment (honzajavorek) [inline juniorguru_chick/lib/threads.py:26]: 👍

File: tests/test_lib_thread_names.py [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+def parse_message_brackets(message, expected_name):
```
Comment (honzajavorek) [inline tests/test_lib_thread_names.py:5]:
```suggestion
def test_parse_message_brackets(message, expected_name):
```

**Comment by dariagrudzien** [comment]
@honzajavorek Thank you for the great tips! 
One thing for future—my intention here wasn't to get a review of the code, just to get unblocked (`Please disregard the actual meat of the thing.`). This is to avoid you wasting time in case I throw something out. If this should happen again and I'd like to show you work in progress which in my mind is not ready for review, how should I explain that?

**Comment by honzajavorek** [comment]
@dariagrudzien You did well, I understood perfectly. Initially I did pass over the meat of the thing, but later I found myself having a bit of time and voluntarily decided I might read the rest and drop a few comments in case I see opportunity to help. Hoping that those won't be taken as unsolicited advice 😅

**Comment by dariagrudzien** [comment]
@honzajavorek As long as everyone wins, we're good!

**Comment by dariagrudzien** [comment]
I decided to take a step back and try TDD. Added passing & failing tests.

File: juniorguru_chick/lib/threads.py [hunk 5]
```diff
@@ -23,20 +24,41 @@ async def fetch_starting_message(thread: discord.Thread) -> discord.Message | No
...
+
+    brackets_regex = re.compile(r"""
+        ^\[     # starts with [
+        .*      # any number of any characters
+        [^\s]   # not a whitespace
+        [^\]]   # not a closing bracket
+        \]      # ends with ]
+        """, re.VERBOSE)
+
+    """If the message includes text in square brackets, use that as name for the thread. Otherwise, use the name template."""
```
Comment (honzajavorek) [inline juniorguru_chick/lib/threads.py:37]: This should be right under the function declaration.

File: juniorguru_chick/lib/threads.py [hunk 6]
```diff
@@ -23,20 +24,41 @@ async def fetch_starting_message(thread: discord.Thread) -> discord.Message | No
...
+    if (match:= re.match(brackets_regex, message.content)) is not None:
```
Comment (honzajavorek) [inline juniorguru_chick/lib/threads.py:38]: I thought it works like this:
```suggestion
    if (match:= brackets_regex.match(message.content)) is not None:
```
But if it works, then I don't mind! I learned something new :)

File: juniorguru_chick/lib/threads.py [hunk 7]
```diff
@@ -23,20 +24,41 @@ async def fetch_starting_message(thread: discord.Thread) -> discord.Message | No
...
+        content = match.group()[1:-1]
```
Comment (honzajavorek) [inline juniorguru_chick/lib/threads.py:39]: If you add parentheses to the regex, you'll be able to get it like this:

```python
r'bla bla bla (foo bar) bla'

.group(0) # bla bla bla foo bar bla
.group(1) # foo bar
```

You can even name them:


```python
r'bla bla bla (?P<iloveregexs>foo bar) bla'

.group(0) # bla bla bla foo bar bla
.group(1) # foo bar
.group('iloveregexs') # foo bar
```

File: juniorguru_chick/lib/threads.py [hunk 8]
```diff
@@ -23,20 +24,41 @@ async def fetch_starting_message(thread: discord.Thread) -> discord.Message | No
...
+        content = content.split(",")
```
Comment (honzajavorek) [inline juniorguru_chick/lib/threads.py:40]:
```suggestion
        parts = content.split(",")
```

File: juniorguru_chick/lib/threads.py [hunk 9]
```diff
@@ -23,20 +24,41 @@ async def fetch_starting_message(thread: discord.Thread) -> discord.Message | No
...
+        processed_content = []
+        for word in content:
```
Comment (honzajavorek) [inline juniorguru_chick/lib/threads.py:42]:
```suggestion
        for part in parts:
```

File: juniorguru_chick/lib/threads.py [hunk 10]
```diff
@@ -23,20 +24,41 @@ async def fetch_starting_message(thread: discord.Thread) -> discord.Message | No
...
+        [^\s]   # not a whitespace
+        [^\]]   # not a closing bracket
+        \]      # ends with ]
+        """, re.VERBOSE)
+
+    """If the message includes text in square brackets, use that as name for the thread. Otherwise, use the name template."""
+    if (match:= re.match(brackets_regex, message.content)) is not None:
+        content = match.group()[1:-1]
+        content = content.split(",")
+        processed_content = []
```
Comment (honzajavorek) [inline juniorguru_chick/lib/threads.py:41]:
```suggestion
        words = []
```

File: juniorguru_chick/lib/threads.py [hunk 11]
```diff
@@ -23,20 +24,41 @@ async def fetch_starting_message(thread: discord.Thread) -> discord.Message | No
...
+        for word in content:
+            processed_content.append(word.lstrip().strip())
```
Comment (honzajavorek) [inline juniorguru_chick/lib/threads.py:43]:
```suggestion
            words.append(part.strip())
```

Having `lstrip().strip()` is redundant, so I removed the `lstrip()`.

File: juniorguru_chick/lib/threads.py [hunk 12]
```diff
@@ -23,20 +24,41 @@ async def fetch_starting_message(thread: discord.Thread) -> discord.Message | No
...
+        name = ", ".join(processed_content)
```
Comment (honzajavorek) [inline juniorguru_chick/lib/threads.py:44]:
```suggestion
        name = ", ".join(words)
```

File: tests/test_lib_thread_names.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+def test_name_thread_no_brackets():
+    content = "Hello this is my thread"
+    weekday = datetime.now().weekday()
+    name_template = "{weekday} objev od {author}"
+    expected_name = f"{DAYS[weekday]} objev od Jana"
+    message = Message("Jana", content)
+
+    assert name_thread(message, name_template) == expected_name
+
+# test if brackets are not closed and if there is no text in brackets
```
Comment (honzajavorek) [inline tests/test_lib_thread_names.py:24]: Either this comment is redundant or `test_brackets_used_incorrectly` is not clear. If `test_brackets_used_incorrectly` feels clear to you, we can remove this comment. If `test_brackets_used_incorrectly` doesn't feel sufficient, it's completely okay to name tests like `test_if_brackets_are_not_closed_and_if_there_is_no_text_in_brackets` 😄

**Review by honzajavorek — APPROVED** [review]
As far as the tests pass, I found no blockers! Made a few suggestions, but up to you.

Only now I realized the name is now composed from just the words, but in https://github.com/juniorguru/juniorguru-chick/issues/10 the idea is more like „Past na Honza Javorek: algoritmy“ instead of „Čtvrteční past na Honza Javorek”. I think we can improve it this way as a next step in a separate PR though.

File: juniorguru_chick/lib/threads.py (hunk 6 above)
Comment (dariagrudzien) [reply juniorguru_chick/lib/threads.py:38]: Yeah, that's what I gathered from https://docs.python.org/3/library/re.html#re.match 
But the test pass with your approach, too. Backwards compatibility?

File: juniorguru_chick/lib/threads.py (hunk 7 above)
Comment (dariagrudzien) [reply juniorguru_chick/lib/threads.py:39]: Ah, thank you for explaining this! I looked at the example which you provided previously and went "how do I even read this". My bad, I should have spent more time with it.

**Pull request merged**